winch/codegen/src/isa/x64/masm.rs

```diff
@@ -1670,6 +1670,98 @@ impl Masm for MacroAssembler {
+        dst: WritableReg,
+        size: OperandSize,
+        handle_overflow_kind: HandleOverflowKind,
+    ) -> Result<()> {
+        self.ensure_has_avx()?;
+
+        let op = match handle_overflow_kind {
+            HandleOverflowKind::None => match size {
+                OperandSize::S8 => AvxOpcode::Vpaddb,
+                OperandSize::S16 => AvxOpcode::Vpaddw,
+                OperandSize::S32 => AvxOpcode::Vpaddd,
+                OperandSize::S64 => AvxOpcode::Vpaddq,
+                OperandSize::S128 => bail!(CodeGenError::unexpected_operand_size()),
+            },
+            HandleOverflowKind::SignedSaturating => match size {
+                OperandSize::S8 => AvxOpcode::Vpaddsb,
+                OperandSize::S16 => AvxOpcode::Vpaddsw,
+                _ => bail!(CodeGenError::unexpected_operand_size()),
+            },
+            HandleOverflowKind::UnsignedSaturating => match size {
+                OperandSize::S8 => AvxOpcode::Vpaddusb,
+                OperandSize::S16 => AvxOpcode::Vpaddusw,
+                _ => bail!(CodeGenError::unexpected_operand_size()),
+            },
+        };
+
+        self.asm.xmm_rmi_rvex(op, lhs, rhs, dst);
+
+        Ok(())
+    }
+
+    fn v128_sub(
+        &mut self,
+        lhs: Reg,
+        rhs: Reg,
+        dst: WritableReg,
+        size: OperandSize,
+        handle_overflow_kind: HandleOverflowKind,
+    ) -> Result<()> {
+        self.ensure_has_avx()?;
+
+        let op = match handle_overflow_kind {
+            HandleOverflowKind::None => match size {
+                OperandSize::S8 => AvxOpcode::Vpsubb,
+                OperandSize::S16 => AvxOpcode::Vpsubw,
+                OperandSize::S32 => AvxOpcode::Vpsubd,
+                OperandSize::S64 => AvxOpcode::Vpsubq,
+                OperandSize::S128 => bail!(CodeGenError::unexpected_operand_size()),
+            },
+            HandleOverflowKind::SignedSaturating => match size {
+                OperandSize::S8 => AvxOpcode::Vpsubsb,
+                OperandSize::S16 => AvxOpcode::Vpsubsw,
+                _ => bail!(CodeGenError::unexpected_operand_size()),
+            },
+            HandleOverflowKind::UnsignedSaturating => match size {
+                OperandSize::S8 => AvxOpcode::Vpsubusb,
+                OperandSize::S16 => AvxOpcode::Vpsubusw,
+                _ => bail!(CodeGenError::unexpected_operand_size()),
+            },
+        };
+
+        self.asm.xmm_rmi_rvex(op, lhs, rhs, dst);
+
+        Ok(())
+    }
+
+    fn v128_mul(&mut self, lhs: Reg, rhs: Reg, dst: WritableReg, size: OperandSize) -> Result<()> {
+        let mul_avx = |this: &mut Self, op| {
+            this.ensure_has_avx()?;
+            this.asm.xmm_rmi_rvex(op, lhs, rhs, dst);
+            Ok(())
+        };
+
+        let mul_avx512 = |this: &mut Self, op| {
+            this.ensure_has_avx512vl()?;
+            this.ensure_has_avx512dq()?;
+            this.asm.xmm_rm_rvex3(op, lhs, rhs, dst);
+            Ok(())
```

Comment:
I don't think we should make this a hard requirement, given that our baseline is AVX.

Given that Intel [suggests that there's no penalty on mixing AVX with AVX512 instructions ](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-avx-512-instructions.html), we could emit AVX512 if they are available, however in case they aren't we still need to emit a fallback to avoid bumping our baseline for this operation. For reference: https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/codegen/src/isa/x64/lower.isle#L1121

Reply:
my bad, I misunderstood your DM about what we expected to support. I have ported cranelift fallback implementation.

Reply:
hold on I just foudn a bug

Reply:
Fixed